Mathlib/Topology/Covering.lean

```diff
@@ -190,3 +191,207 @@ protected theorem FiberBundle.isCoveringMap {F : Type*} {E : X → Type*} [Topol
+    map_source' := fun x hx ↦ ⟨hx, trivial⟩
+    map_target' := fun x ⟨hx, _⟩ ↦ by dsimp only; rw [dif_pos hx]; apply (f_inv _ hx).symm ▸ hx,
+    left_inv' := fun e he ↦ ?_,
+    right_inv' := fun x hx ↦ ?_ }
+  · change f e ∈ V at he; dsimp only; simp_rw [dif_pos he]
+    exact inj _ (inv_U _ he) (idx_U e he) (f_inv _ _)
+  · dsimp only; rw [dif_pos hx.1]
+    refine Prod.ext (f_inv _ hx.1) ?_
+    dsimp only; rw [dif_pos ((f_inv _ hx.1).symm ▸ hx.1)]
+    by_contra h; exact (disjoint h).le_bot ⟨idx_U _ _, inv_U _ _⟩
+  have open_preim : ∀ {W}, W ⊆ V → IsOpen W → IsOpen (f ⁻¹' W)
+  · intro W hWV hoW
+    convert isOpen_iUnion (fun i ↦ (open_iff i hWV).mp hoW)
+    rw [← Set.inter_iUnion, eq_comm, Set.inter_eq_left]
+    exact (Set.preimage_mono hWV).trans exhaustive'
+  have open_source : IsOpen F.source := open_preim subset_rfl open_V
+  have cont_f : ContinuousOn f F.source := (continuousOn_open_iff open_source).mpr
+    fun W open_W ↦ open_preim (V.inter_subset_left W) (open_V.inter open_W)
+  refine
+  { toLocalEquiv := F,
+    open_source := open_source,
+    open_target := open_V.prod isOpen_univ,
+    continuous_toFun := cont_f.prod <| ContinuousAt.continuousOn fun e he ↦
+      continuous_const (b := idx e he) |>.continuousAt.congr <| mem_nhds_iff.mpr
+        ⟨U (idx e he) ∩ F.source, fun e' he' ↦ ?_, ?_, idx_U e he, he⟩
+    continuous_invFun := continuousOn_prod_of_discrete_right.mpr fun i ↦ ?_,
+    baseSet := V,
+    open_baseSet := open_V,
+    source_eq := rfl,
+    target_eq := rfl,
+    proj_toFun := fun _ _ ↦ rfl }
+  · by_contra h; apply (disjoint h).le_bot
+    dsimp only; erw [dif_pos he'.2]; exact ⟨he'.1, idx_U _ _⟩
+  · rwa [Set.inter_comm, ← open_iff _ subset_rfl]
+  · simp_rw [Set.prod_mk_mem_set_prod_eq, Set.mem_univ, and_true]
+    refine (continuousOn_open_iff open_V).mpr fun W open_W ↦ ?_
+    rw [open_iff i (V.inter_subset_left _)]
+    convert ((open_iff i subset_rfl).mp open_V).inter open_W using 1
+    ext e; refine and_right_comm.trans (and_congr_right fun ⟨hV, hU⟩ ↦ ?_)
+    rw [Set.mem_preimage, dif_pos hV, inj i (inv_U i _) hU (f_inv i _)]
+
+namespace QuotientMap
+
+open Topology
+
+variable {G} [t : TopologicalSpace G] [d : DiscreteTopology G]
+  [Group G] [MulAction G E] [ContinuousConstSMul G E]
+  (hf : QuotientMap f) (hfG : ∀ {e₁ e₂}, f e₁ = f e₂ ↔ e₁ ∈ MulAction.orbit G e₂)
+
+/-- If a group `G` acts on a space `E` and `U` is an open subset disjoint from all other
+  `G`-translates of itself, and `p` is a quotient map by this action, then `p` admits a
+  `Trivialization` over the base set `p(U)`. -/
+@[to_additive "If a group `G` acts on a space `E` and `U` is an open subset disjoint from all
+  other `G`-translates of itself, and `p` is a quotient map by this action, then `p` admits a
+  `Trivialization` over the base set `p(U)`."]
+noncomputable def trivialization_of_smul_disjoint (U : Set E) (open_U : IsOpen U)
+    (disjoint : ∀ g : G, (g • ·) '' U ∩ U ≠ ∅ → g = 1) : Trivialization G f := by
```

Comment:
In this case I think it's actually easier to build the trivialization rather than working with unions. Indeed the trivialization is just `(g, x) ↦ g • x`, which is continuous and open (see #7809), and you just have to prove that it's a bijection between `G • U` and `f ⁻¹' (f '' U)`.

Reply:
Also, why not using pointwise notation? I think you'd have better API available.

Reply:
When you say `(g, x) ↦ g • x`, the first `x` lies in X (downstairs) while the second `x` lies in E (upstairs), but the topology on E and X are not a priori "the same" (local homeomorphism) and the assumption is only that `f : E → X` is a quotient map. So I have to resort to the definition to show the continuity in the direction `G × X → E` (suffices to restrict to fixed `g : G` by my previous PR). I think most of the gymnastics in this proof is showing that preimage of an open set `W` in `E` is `f (g⁻¹ (W ∩ U))`, and that its `f`-preimage is open by writing it as the union of `G`-translates of `W ∩ U`.

This was originally [an even longer proof](https://gist.github.com/alreadydone/40333872e072abc8125dd95d20f07f2c#file-properlydiscontinuous_coveringmap-lean-L38-L104) but I decided to extract `IsOpen.trivialization_discrete` (which can be applied to (maybe purely hypothetical) situations without a group action); a downsize is that the length of these two proofs combined is 15 lines longer than the original one.

I used `(g • ·) '' U` because this is what's used in [ProperlyDiscontinuousSMul](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Algebra/ConstMulAction.html#ProperlyDiscontinuousSMul). I haven't looked much into the Pointwise API, but I did use one lemmas that uses pointwise notation, [singleton_mul_mem_nhds_of_nhds_one](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Algebra/Group/Basic.html#singleton_mul_mem_nhds_of_nhds_one). If you have golfs that's certainly welcome!